web/src/upload_widget.ts

```diff
@@ -25,31 +41,122 @@ export function build_widget(
     $upload_button: JQuery,
     $preview_text?: JQuery,
     $preview_image?: JQuery,
+    $select_button?: JQuery,
+    $all_elements_to_hide?: JQuery,
+    $placeholder_icon?: JQuery,
     max_file_upload_size = default_max_file_size,
 ): {clear(): void; close(): void} {
     function accept(file: File): void {
         $file_name_field.text(file.name);
+        $file_name_field.attr("title", file.name);
         $input_error.hide();
+        $all_elements_to_hide?.hide();
         $clear_button.show();
+        $select_button?.show();
         $upload_button.hide();
+
         if ($preview_text !== undefined && $preview_image !== undefined) {
-            const image_blob = URL.createObjectURL(file);
-            $preview_image.attr("src", image_blob);
+            $placeholder_icon?.hide();
+            $preview_image.hide();
             $preview_text.show();
+
+            // opening uppy
+            uppy = new Uppy({
+                id: "emoji-uppy",
+                allowMultipleUploadBatches: false,
+                restrictions: {
+                    allowedFileTypes: supported_types,
+                    maxNumberOfFiles: 1,
+                    maxFileSize: max_file_upload_size * 1024 * 1024,
+                },
+            })
+                .use(Dashboard, {
+                    id: "Dashboard",
+                    inline: true,
+                    hideCancelButton: true,
+                    hideUploadButton: true,
+                    height: 450,
+                    target: "#" + String($preview_text.attr("id")),
+                    autoOpenFileEditor: true,
+                    proudlyDisplayPoweredByUppy: false,
+                    theme: settings_data.using_dark_theme() ? "dark" : "light",
+                })
+                .use(ImageEditor, {
+                    id: "ImageEditor",
+                    target: Dashboard,
+                    quality: 1,
+                    cropperOptions: {
+                        viewMode: 0,
+                        background: true,
+                        autoCropArea: 1,
+                        aspectRatio: 1,
+                        responsive: true,
+                        croppedCanvasOptions: {},
+                    },
+                    actions: {
+                        revert: true,
+                        rotate: true,
+                        granularRotate: true,
+                        flip: true,
+                        zoomIn: true,
+                        zoomOut: true,
+                        cropSquare: true,
+                        cropWidescreen: false,
+                        cropWidescreenVertical: false,
+                    },
+                })
```

Comment:
Is there a way to avoid this large block of duplicated uppy configuration code? You seem to do this twice; once here and once in `accept` below.

Reply:
I have addressed this issue in the updated PR and have reduced the redundancy.